pkg/hive/cell/structured.go

```diff
@@ -0,0 +1,384 @@
+}
+
+// JSON serializable status update tree node.
+type StatusNode struct {
+	ID          string        `json:"id"`
+	Level       Level         `json:"level,omitempty"`
+	Name        string        `json:"name"`
+	Message     string        `json:"message,omitempty"`
+	Timestamp   time.Time     `json:"timestamp"`
+	Count       int           `json:"count"`
+	SubStatuses []*StatusNode `json:"sub_statuses,omitempty"`
+}
+
+func (s *StatusNode) allOk() bool {
+	return s.Level == StatusOK
+}
+
+func (s *StatusNode) writeTo(w io.Writer, d int) {
+	if len(s.SubStatuses) == 0 {
+		fmt.Fprintf(w, "%s%s\t%s\t%s\t%s ago\t(x%d)\n", strings.Repeat("\t", d), s.Name, s.Level, s.Message, time.Since(s.Timestamp), s.Count)
+	} else {
+		fmt.Fprintf(w, "%s%s\n", strings.Repeat("\t", d), s.Name)
+		for _, ss := range s.SubStatuses {
+			ss.writeTo(w, d+1)
+		}
+	}
+}
+
+func (s *StatusNode) toString(ident int) string {
+	buf := bytes.NewBuffer(nil)
+	w := tabwriter.NewWriter(buf, 0, 0, 1, ' ', 0)
+	s.writeTo(w, ident)
+	w.Flush()
+	return buf.String()
+}
+
+func (s *StatusNode) String() string {
+	return s.toString(0)
+}
+
+// If all leaves are Ok, then we are ok
+func (s *subreporterBase) getStatusTreeLocked(nid string) *StatusNode {
+	if children, ok := s.idToChildren[nid]; ok {
+		rn := s.nodes[nid]
+		n := &StatusNode{
+			ID:        nid,
+			Message:   rn.Message,
+			Name:      rn.name,
+			Timestamp: rn.Timestamp,
+			Count:     rn.count,
+		}
+		allok := true
+		childIDs := maps.Keys(children)
+		sort.Strings(childIDs)
+		for _, child := range childIDs {
+			cn := s.getStatusTreeLocked(child)
+			if cn == nil {
+				log.Errorf("failed to get status for node %s", child)
+				continue
+			}
+			n.SubStatuses = append(n.SubStatuses, cn)
+			if !cn.allOk() {
+				allok = false
+			}
+		}
+		// If this is not a leaf and all children are ok then report ok.
+		// case 1: Non-reporter, has no children, should be ok?
+		// case 2: Non-reporter, has children, defer down to children.
+		if rn.isReporter {
+			n.Level = rn.Level
+		} else {
+			if allok {
+				n.Level = StatusOK
+			} else {
+				n.Level = StatusDegraded
+			}
+		}
+
+		return n
+	}
+	return nil
+}
+
+type node struct {
+	id         string
+	name       string
+	parentID   string
+	isReporter bool
+	count      int
+	Update
+}
+
+// subReporter represents both reporter "leaf" nodes and intermediate
+// "scope" nodes.
+// subReporter only has a pointer to the base, thus copying a subReporter
+// by value yields the same "reporter".
+type subReporter struct {
+	base    *subreporterBase
+	realize func()
+	id      string
+	name    string
+}
+
+func (s *subReporter) OK(message string) {
+	if err := s.base.setStatus(s.id, StatusOK, message); err != nil {
+		log.WithError(err).Warnf("ok on subreporter that has been stopped")
+		return
+	}
+	s.realize()
+}
+
+func (s *subReporter) Degraded(message string, err error) {
+	var m string
+	// TODO: For now, we just append the error to the message, this won't be ideal
```

Comment:
Can we address this TODO now? If not, perhaps doing a follow-up issue would make sense.

Reply:
Yeah I'm kind of kicking this can down the road a bit 😅 

While working on this I went back and fourth on removing the error from the `Degraded` function.  I ended up keeping it to keep the PR smaller and since I'm still not entirely sure what the HealthReporter API should ultimately look like.

I think eventually we'll want to replace the passed status values with some kind of interface like:

```go
type Report interface {
    String() string
    JSON() []byte
    Errs() []error
}
```
(This will also be important since currently we send the status update as raw text so on the receiving end we have no good way to decode the structure).

I'll address this in a follow up after these changes are merged.